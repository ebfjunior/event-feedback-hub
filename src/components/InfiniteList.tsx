/* eslint-disable react-hooks/exhaustive-deps */
'use client';

import { useEffect, useRef } from 'react';
import { cn } from '@/lib/utils';

export type InfiniteListProps = {
  items: React.ReactNode[];
  hasMore: boolean;
  isLoading: boolean;
  onLoadMore: () => void;
  className?: string;
};

export function InfiniteList({ items, hasMore, isLoading, onLoadMore, className }: InfiniteListProps) {
  const sentinelRef = useRef<HTMLDivElement | null>(null);

  useEffect(() => {
    if (!hasMore) return;
    const el = sentinelRef.current;
    if (!el) return;
    const io = new IntersectionObserver((entries) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) onLoadMore();
      });
    });
    io.observe(el);
    return () => io.disconnect();
  }, [hasMore]);

  return (
    <div className={cn('space-y-3', className)}>
      {items}
      {isLoading && <div className="text-center text-sm text-muted-foreground">Loadingâ€¦</div>}
      {!hasMore && !isLoading && (
        <div className="text-center text-sm text-muted-foreground">End of the list</div>
      )}
      <div ref={sentinelRef} />
    </div>
  );
}
