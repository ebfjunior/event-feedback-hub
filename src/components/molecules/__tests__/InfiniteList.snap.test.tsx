import React from 'react';
import { render } from '@testing-library/react';
import { InfiniteList } from '../InfiniteList';

describe('InfiniteList (snapshot)', () => {
	it('renders loading and sentinel states', () => {
		const { container: c1 } = render(
			<InfiniteList items={[<div key="1">A</div>, <div key="2">B</div>]} hasMore isLoading onLoadMore={() => {}} />,
		);
		expect(c1.firstChild).toMatchSnapshot();
		const { container: c2 } = render(
			<InfiniteList items={[]} hasMore={false} isLoading={false} onLoadMore={() => {}} />,
		);
		expect(c2.firstChild).toMatchSnapshot();
	});
});


