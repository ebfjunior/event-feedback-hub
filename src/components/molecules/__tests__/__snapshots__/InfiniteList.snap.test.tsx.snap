// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfiniteList (snapshot) > renders loading and sentinel states 1`] = `
<div
  class="space-y-3"
>
  <div>
    A
  </div>
  <div>
    B
  </div>
  <div
    class="text-center text-sm text-muted-foreground"
  >
    Loadingâ€¦
  </div>
  <div />
</div>
`;

exports[`InfiniteList (snapshot) > renders loading and sentinel states 2`] = `
<div
  class="space-y-3"
>
  <div
    class="text-center text-sm text-muted-foreground"
  >
    End of the list
  </div>
  <div />
</div>
`;
